astropy/modeling/tests/test_quantities_evaluation.py

```diff
@@ -0,0 +1,538 @@
+
+"""
+Tests that relate to evaluating models with quantity parameters
+"""
+
+from __future__ import (absolute_import, unicode_literals, division,
+                        print_function)
+
+from numpy.testing import assert_allclose
+
+
+from ..core import Fittable1DModel
+from ..parameters import Parameter
+from ..models import Gaussian1D
+from ... import units as u
+from ...units import UnitsError
+from ...tests.helper import pytest, assert_quantity_allclose
+
+
+# We start off by taking some simple cases where the units are defined by
+# whatever the model is initialized with, and we check that the model evaluation
+# returns quantities.
+
+
+@pytest.mark.xfail
+def test_evaluate_with_quantities():
+    """
+    Test evaluation of a single model with Quantity parameters that do
+    not explicitly require units.
+    """
+
+    # We create two models here - one with quantities, and one without. The one
+    # without is used to create the reference values for comparison.
+
+    g = Gaussian1D(1, 1, 0.1)
+    gq = Gaussian1D(1 * u.J, 1 * u.m, 0.1 * u.m)
+
+    # We first check that calling the Gaussian with quantities returns the
+    # expected result
+    assert_quantity_allclose(gq(1 * u.m), g(1) * u.J)
+
+    # Units have to be specified for the Gaussian with quantities - if not, an
+    # error is raised
+    with pytest.raises(UnitsError) as exc:
+        gq(1)
+    assert exc.value.args[0] == ("Units of input 'x', (dimensionless), could not be "
+                                 "converted to required input units of m (length)")
+
+    # However, zero is a special case
+    assert_quantity_allclose(gq(0), g(0) * u.J)
+
+    # We can also evaluate models with equivalent units
+    assert_allclose(gq(0.0005 * u.km).value, g(0.5))
+
+    # But not with incompatible units
+    with pytest.raises(UnitsError) as exc:
+        gq(3 * u.s)
+    assert exc.value.args[0] == ("Units of input 'x', s (time), could not be "
+                                 "converted to required input units of m (length)")
+
+    # We also can't evaluate the model without quantities with a quantity
+    with pytest.raises(UnitsError) as exc:
+        g(3 * u.m)
+    assert exc.value.args[0] == ("Input in m (length), could not be converted "
+                                 "to required to dimensionless input")
+
+
+@pytest.mark.xfail
+def test_evaluate_with_quantities_and_equivalencies():
+    """
+    We now make sure that equivalencies are correctly taken into account
+    """
+
+    g = Gaussian1D(1 * u.Jy, 10 * u.nm, 2 * u.nm)
+
+    # We haven't set the equivalencies yet, so this won't work
+    with pytest.raises(UnitsError) as exc:
+        g(30 * u.PHz)
+    assert exc.value.args[0] == ("Units of input 'x', PHz (frequency), could "
+                                 "not be converted to required input units of "
+                                 "nm (length)")
+
+    g.mean.equivalencies = u.spectral()
+    g.stddev.equivalencies = u.spectral()
+
```

Comment:
Could this be punted to the outside -- i.e., assume that you normally define models with quantities such that they make sense without equivalencies, and if you want more, you do your fitting with equivalencies enabled? Alternatively, with the introduction of functional units like `mag`, the `unit` machinery has gained the ability for units to carry equivalencies. It may be possible to use that here too, i.e., one would define the Gaussian model not using, say, `u.Angstrom`, but a new unit that is Angstrom too but has equivalencies set.